File: isegoria/src/main/java/com/eulersbridge/isegoria/util/transformation/RoundedCornersTransformation.java
package com.eulersbridge.isegoria.util.transformation;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.RectF;
import android.support.annotation.NonNull;

import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation;
import com.eulersbridge.isegoria.util.Constants;

import java.security.MessageDigest;

public class RoundedCornersTransformation extends BitmapTransformation {
    private static final String ID = String.format("%s.RoundedCornersTransformation", Constants.APP_ID);
    private static final byte[] ID_BYTES = ID.getBytes(CHARSET);

    // Radius of rounded corners in DP
    private final int cornerRadius;

    private static float screenDensity = Resources.getSystem().getDisplayMetrics().density;

    public RoundedCornersTransformation() {
        this(0);
    }

    @SuppressWarnings("WeakerAccess")
    public RoundedCornersTransformation(int cornerRadius) {
        this.cornerRadius = cornerRadius;
    }

    @Override
    protected Bitmap transform(@NonNull BitmapPool pool, @NonNull Bitmap toTransform, int outWidth, int outHeight) {
        Bitmap bitmap = pool.get(outWidth, outHeight, Bitmap.Config.ARGB_8888);

        Canvas canvas = new Canvas(bitmap);

        final RectF rect = new RectF(0, 0, outWidth, outHeight);

        Paint paint = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.DITHER_FLAG);
        paint.setColor(Color.BLACK);

        int cornerRadiusPx = Math.round((float)cornerRadius * screenDensity);
        canvas.drawRoundRect(rect, cornerRadiusPx, cornerRadiusPx, paint);

        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        canvas.drawBitmap(toTransform, 0, 0, paint);

        return bitmap;
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof RoundedCornersTransformation;
    }

    @Override
    public int hashCode() {
        return ID.hashCode();
    }

    @Override
    public void updateDiskCacheKey(@NonNull MessageDigest messageDigest) {
        messageDigest.update(ID_BYTES);
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/util/transformation/RoundedCornersTransformation.kt
package com.eulersbridge.isegoria.util.transformation

import android.graphics.*
import android.support.annotation.IntRange
import android.support.annotation.Px
import androidx.graphics.applyCanvas
import androidx.graphics.toXfermode
import com.bumptech.glide.load.Key
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation
import com.eulersbridge.isegoria.APP_ID
import java.security.MessageDigest

class RoundedCornersTransformation(
    // Radius of rounded corners in DP
    @IntRange(from = 1) private val cornerRadius: Int
) : BitmapTransformation() {

    companion object {
        private const val ID = "$APP_ID.RoundedCornersTransformation"
        private val ID_BYTES = ID.toByteArray(Key.CHARSET)

        var screenDensity = 2.0f
    }

    constructor() : this(1)

    override fun transform(
        pool: BitmapPool,
        toTransform: Bitmap,
        outWidth: Int,
        outHeight: Int
    ): Bitmap {
        val bitmap = pool.get(outWidth, outHeight, Bitmap.Config.ARGB_8888)

        return bitmap.applyCanvas {
            val rect = RectF(0f, 0f, outWidth.toFloat(), outHeight.toFloat())
            @Px val cornerRadiusPx = Math.round(cornerRadius.toFloat() * screenDensity)

            val paint = Paint(Paint.ANTI_ALIAS_FLAG or Paint.DITHER_FLAG)
            paint.color = Color.BLACK

            drawRoundRect(rect, cornerRadiusPx.toFloat(), cornerRadiusPx.toFloat(), paint)

            paint.xfermode = PorterDuff.Mode.SRC_IN.toXfermode()

            drawBitmap(toTransform, 0f, 0f, paint)
        }
    }

    override fun equals(other: Any?) = other is RoundedCornersTransformation

    override fun hashCode() = ID.hashCode()

    override fun updateDiskCacheKey(messageDigest: MessageDigest) = messageDigest.update(ID_BYTES)
}
